lib/buffer.js

```diff
@@ -392,17 +392,25 @@ Buffer.prototype.compare = function compare(b) {
 };
 
 
-Buffer.prototype.indexOf = function indexOf(val, byteOffset) {
+Buffer.prototype.indexOf = function indexOf(val, byteOffset, encoding) {
   if (byteOffset > 0x7fffffff)
     byteOffset = 0x7fffffff;
   else if (byteOffset < -0x80000000)
     byteOffset = -0x80000000;
   byteOffset >>= 0;
 
-  if (typeof val === 'string')
-    return binding.indexOfString(this, val, byteOffset);
+  if (typeof val === 'string') {
+    if (encoding === undefined ||
+        encoding === 'ucs2' ||
+        encoding === 'utf16le' ||
+        encoding === 'utf8') {
+      return binding.indexOfString(this, val, byteOffset, encoding);
+    }
+    return binding.indexOfBuffer(
+        this, Buffer(val, encoding), byteOffset, encoding);
```

Comment:
Yeah. Here's a naive impl that I had worked on: https://gist.github.com/trevnorris/d39ff9bc359d2e406faf#file-simple-string-search-cc-L40-L45

You'll see that I'm allocating memory for the new string, but this is still faster than calling `new Buffer()`. As a performance optimization, and assuming most strings are < 1KB, could also do `char needle[1024]` and use that instead if the string fits.